### ISSUE TITLE
Fix Warning that appears at VS launch after installing HotCommands

### ISSUE THREAD

**Issue by justcla** [justcla/HotCommands#97]
Error occurs when VS opens.

![image](https://user-images.githubusercontent.com/17131343/221439776-444a957d-8d0f-446a-8ba0-90d2d6170fbe.png)

`Still unable to load MEF component DLL: Could not load file or assembly 'Microsoft.VisualStudio.Threading, Version=17.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)          C:\Users\Justin\AppData\Local\Microsoft\VisualStudio\17.0_6b28b076\Extensions\pq0u3rou.bbr\HotCommands.dll`

<hr>

`SetSite failed for package [VSShortcutsManagerPackage]Source: 'KeyboardShortcutsManager' Description: Object reference not set to an instance of an object. System.NullReferenceException: Object reference not set to an instance of an object. at VSShortcutsManager.ShortcutsScanner.ScanForNewShortcutsDefs() in C:\Projects\VSShortcutsManager\VSShortcutsManager\ExtensionScanner\ShortcutsScanner.cs:line 250 at VSShortcutsManager.ShortcutsScanner.ScanForAllExtensionShortcuts() in C:\Projects\VSShortcutsManager\VSShortcutsManager\ExtensionScanner\ShortcutsScanner.cs:line 193 at VSShortcutsManager.VSShortcutsManager..ctor(Package package) in C:\Projects\VSShortcutsManager\VSShortcutsManager\VSShortcutsManager.cs:line 137 at VSShortcutsManager.VSShortcutsManagerPackage.<InitializeAsync>d__7.MoveNext() in C:\Projects\VSShortcutsManager\VSShortcutsManager\VSShortcutsManagerPackage.cs:line 51 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.VisualStudio.Shell.AsyncPackage.<>c__DisplayClass21_0.<<Microsoft-VisualStudio-Shell-Interop-IAsyncLoadablePackageInitialize-Initialize>b__1>d.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at Microsoft.VisualStudio.Services.VsTask.RethrowException(AggregateException e) at Microsoft.VisualStudio.Services.VsTask.InternalGetResult(Boolean ignoreUIThreadCheck)`

**Comment by justcla** [comment]
Most likely related to the VSShortcutsManager extension ([Keyboard Shortcuts Manager](https://marketplace.visualstudio.com/items?itemName=JustinClareburtMSFT.VSShortcutsManager)), which is embedded with HotCommand so that it loads the shortctus after instal (triggered by the ShortcutsScanner).

**Comment by justcla** [comment]
@cfognom A new challenge?

This would involve fixing the bug in the Keyboard Shortcuts Manager.

This is the GitHub project. https://github.com/justcla/VSShortcutsManager

**Comment by justcla** [comment]
Not related to the VSShortcutsScanner. Same error thrown when that extension removed. Thrown on HotCommands.

`Could not load file or assembly 'Microsoft.VisualStudio.Threading, Version=17.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.`

**Comment by cfognom** [comment]
Did you get this warning always? I only started getting this after v 4.3.0. That's why I thought It might be related to my fiddling with the references.

**Comment by justcla** [comment]
Yes, I think it's because it's looking for v17.5 references, and I'm still on 17.4.
For safety, I'm downgrading all references to 17.0

**Comment by justcla** [comment]
I got it working. But broke it when I called MoveMemberDown. It's tricky to get all the versions matching.

**Comment by justcla** [comment]
I put up my working code in a branch: [fix-references](https://github.com/justcla/HotCommands/tree/fix-references)
It mostly works, I think.

**Comment by tiesmaster** [comment]
Same here, `4.3.1` is also still failing, so I'll revert back to the working version. If you need any help testing/validating, l'm available.

**Comment by justcla** [comment]
@tiesmaster Please help. We updated the references and some of them are too new for people with older installations.
Trying to get everything baselined to v17.0.x libraries.

**Comment by cfognom** [comment]
> I put up my working code in a branch: [fix-references](https://github.com/justcla/HotCommands/tree/fix-references) It mostly works, I think.

I get no warning when I build and debug v4.3.0 or later, including the 'fix-references' branch. My guess is that: As long as the one who is publishing the build has no warning, the users will not get it either. So, If that branch made it work for you, we should be good.

**Comment by justcla** [comment]
> As long as the one who is publishing the build has no warning, the users will not get it either.

This is definitely not true. In fact, it's a trap. "Works on my machine" said the developer - because their machine has the v17.5 version of Microsoft.VisualStudio.Threading. But a user who is on an older version of VS (like me) will not have that version of the library.

I'm currently running with v17.4 libraries, so I can build and run with v17.4, but references to 17.5 break at runtime when I run them - because the files are missing.

The only way I know I can fix this bug is by avoiding the upgrade to v17.5. ;-)

But I generally try to use the oldest possible libraries to ensure the greatest possible reach for users.
Hence why I am trying to target v17.0 libraries, instead of v17.5.

**Comment by cfognom** [comment]
I think you misunderstand me. When I say:

> As long as the one who is publishing the build has no warning, the users will not get it either.

I mean the users of the built extension, not users of the HotCommands repo.

I'm not saying just because the repo has no errors for me, will it have no errors for other people using the repo. It very likely will, just like how I initially couldn't build the repo, but you could.

**Comment by justcla** [comment]
Yes, that's how I inferred it.
Does my response make sense?

What I'm saying is that whether you get errors or not, users on older versions will get runtime errors if you build it with packages they don't have on their machine.

**Comment by cfognom** [comment]
That depends on whether the packages are included in the built extension or just referenced from it. Which one it is I don't know. You probably know this better. But I was under the assumption that the packages were included in the build. Since why otherwise would v4.3.0 work when I build and run on my machine. But not when I run v4.3.0 built by you?

**Comment by cfognom** [comment]
If all else fails, you have the option to revert my commit to the .csproj file and go back to the old way the packages was referenced. https://github.com/justcla/HotCommands/commit/f59388acc6f96c852c99b75df808fe85cd10f384

**Comment by justcla** [comment]
Some are compile-time dependencies (needed when building). Some are runtime dependencies (needed only when they feature is triggered). Some are both.
It's hard to know which is which. I don't even know why half those dependencies are there.

**Comment by tiesmaster** [comment]
@justcla I'm not sure how I can help with this. These assembly hell issues are always difficult to troubleshoot, and I'm still trying to clear out the confusion between you and @cfognom 😬 and I don't want to add to the confusion, which might only cause it longer to fix this. But in case you do want any help on that topic, I would need to dive in there, and investigate some things myself, like checkout the VSIX packages, and such, and see what kind reference ILSpy, and such tells us. Let me know if you appreciate that.

**Comment by justcla** [comment]
@tiesmaster It's OK. I think I can manage it. Just need to undo one commit. I've made a version that builds (on my machine) with the old references, which get auto-upgraded by users on new versions.

**Comment by justcla** [comment]
So it turned out to just be one odd reference - Microsoft.VisualStudio.Threading was referencing v17.5.22 while everything else was referencing lower versions (v17.4.x).
I downgraded Microsoft.VisualStudio.Threading to v17.4.33 so now it works on versions of VS that are 17.4.3 and higher.

I expect it will still cause errors for users of older versions of VS. But...this will do for now.
If there is a user who cannot upgrade from the base 17.0 (and that happens for various reasons), we can revisit this and downgrade all the references to v17.0.x.

New version 4.3.2 is available on the Marketplace. Please test it and see if it works for you. Thanks.

**Comment by cfognom** [comment]
Yep, the warning is gone. Great work!

**Comment by tiesmaster** [comment]
🤯 Jeez, bizarre. Thank you @justcla for fixing this 🎉
